prow/bugzilla/client.go

```diff
@@ -214,37 +284,95 @@ func getRootForClone(c Client, bug *Bug) (*Bug, error) {
 
 // GetRootForClone returns the original bug.
 func (c *client) GetRootForClone(bug *Bug) (*Bug, error) {
-	return getRootForClone(c, bug)
+	bugCache := NewBugDetailsCache()
+	return getRootForClone(c, bug, bugCache)
 }
 
 // GetAllClones returns all the clones of the bug including itself
 // Differs from GetClones as GetClones only gets the child clones which are one level lower
-func (c *client) GetAllClones(bug *Bug) ([]*Bug, error) {
-	return getAllClones(c, bug)
+func (c *client) GetAllClones(bugID int) (clones []*Bug, bug *Bug, root *Bug, err error) {
+	bugCache := NewBugDetailsCache()
+	return getAllClones(c, bugID, bugCache)
+}
+
+func (c *client) GetFromCache(key int) ([]int, bool) {
+	return c.clonesCache.Get(key)
 }
-func getAllClones(c Client, bug *Bug) ([]*Bug, error) {
-	root, err := getRootForClone(c, bug)
+
+func (c *client) SetInCache(key int, value []int) {
+	c.clonesCache.Set(key, value)
+}
+
+func getAllClones(c Client, bugID int, bugCache *BugDetailsCache) ([]*Bug, *Bug, *Bug, error) {
+
+	clonesID, ok := c.GetFromCache(bugID)
+	var bug *Bug
+	g := new(errgroup.Group)
+	g.Go(func() error {
+		var err error
+		bug, err = c.GetBug(bugID)
+		if err != nil {
+			return err
+		}
+		bugCache.Set(bugID, bug)
+		return nil
+	})
+	if ok {
+		for _, cloneID := range clonesID {
+			cloneID := cloneID
+			g.Go(func() error {
+				var clone *Bug
+				clone, ok = bugCache.Get(cloneID)
+				if !ok {
+					var err error
+					clone, err = c.GetBug(cloneID)
+					if err != nil {
+						return err
+					}
+				}
+				bugCache.Set(clone.ID, clone)
+				return nil
+			})
+		}
+	}
+	if err := g.Wait(); err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get bug details: %v", err)
+	}
+
+	root, err := getRootForClone(c, bug, bugCache)
 	if err != nil {
-		return nil, err
+		return nil, nil, nil, err
 	}
-	clones, err := getRecursiveClones(c, root)
+	recursedClones, err := getRecursiveClones(c, root, bugCache)
 	if err != nil {
-		return nil, err
+		return nil, nil, nil, err
+	}
+	// Populate the main cache with relevant cloneIDs
+	// Remember to use a mutex to ensure thread safety
+	// Do not move this section after the section where the original bug is removed
+	cloneSet := sets.NewInt()
+	for bugID := range bugCache.cache {
+		cloneSet.Insert(bugID)
+	}
+	for _, clone := range recursedClones {
+		cloneSet.Delete(clone.ID)
+		c.SetInCache(clone.ID, cloneSet.List())
+		cloneSet.Insert(clone.ID)
```

Comment:
Er. What is happening here? We delete an ID from the set, then set something in the cache, then insert it again, so that we can multiple times overwrite the cache with some subset of the `cloneSet`... That does not make sense to me, can you explain why we are doing this?

Reply:
Yup, the intention here is to set the cache to remember all the associations so that we can parallelize all the calls the next time. ie if bug1, bug2 and bug3 are clones of bug4 then bug2, bug3 and bug4 are also clones of bug1 and so on.
That is why I'm setting it in this manner.